specs/_features/peerdas/polynomial-commitments-sampling.md

```diff
@@ -0,0 +1,520 @@
+    """
+    Negative of coefficient form polynomial ``a``
+    """
+    return [(BLS_MODULUS - x) % BLS_MODULUS for x in a]
+```
+
+#### `multiply_polynomialcoeff`
+
+```python
+def multiply_polynomialcoeff(a: PolynomialCoeff, b: PolynomialCoeff) -> PolynomialCoeff:
+    """
+    Multiplies the coefficient form polynomials ``a`` and ``b``
+    """
+    r = [0]
+    for power, coef in enumerate(a):
+        summand = [0] * power + [int(coef) * int(x) % BLS_MODULUS for x in b]
+        r = add_polynomialcoeff(r, summand)
+    return r
+```
+#### `divide_polynomialcoeff`
+
+```python
+def divide_polynomialcoeff(a: PolynomialCoeff, b: PolynomialCoeff) -> PolynomialCoeff:
+    """
+    Long polynomial division for two coefficient form polynomials ``a`` and ``b``
+    """
+    a = [x for x in a]
+    o = []
+    apos = len(a) - 1
+    bpos = len(b) - 1
+    diff = apos - bpos
+    while diff >= 0:
+        quot = div(a[apos], b[bpos])
+        o.insert(0, quot)
+        for i in range(bpos, -1, -1):
+            a[diff + i] = (int(a[diff + i]) - int(b[i]) * int(quot)) % BLS_MODULUS
+        apos -= 1
+        diff -= 1
+    return [x % BLS_MODULUS for x in o]
+```
+
+#### `shift_polynomialcoeff`
+
+```python
+def shift_polynomialcoeff(polynomial_coeff: PolynomialCoeff, factor: BLSFieldElement) -> PolynomialCoeff:
+    """
+    Shift the evaluation of a polynomial in coefficient form by factor.
+    This results in a new polynomial g(x) = f(factor * x)
+    """
+    factor_power = 1
+    inv_factor = pow(int(factor), BLS_MODULUS - 2, BLS_MODULUS)
+    o = []
+    for p in polynomial_coeff:
+        o.append(int(p) * factor_power % BLS_MODULUS)
+        factor_power = factor_power * inv_factor % BLS_MODULUS
+    return o
+```
+
+#### `interpolate_polynomialcoeff`
+
+```python
+def interpolate_polynomialcoeff(xs: Sequence[BLSFieldElement], ys: Sequence[BLSFieldElement]) -> PolynomialCoeff:
+    """
+    Lagrange interpolation: Finds the lowest degree polynomial that takes the value ``ys[i]`` at ``x[i]``
+    for all i.
+    Outputs a coefficient form polynomial. Leading coefficients may be zero.
+    """
+    assert len(xs) == len(ys)
+    r = [0]
+
+    for i in range(len(xs)):
+        summand = [ys[i]]
+        for j in range(len(ys)):
+            if j != i:
+                weight_adjustment = bls_modular_inverse(int(xs[i]) - int(xs[j]))
+                summand = multiply_polynomialcoeff(
+                    summand, [(- int(weight_adjustment) * int(xs[j])) % BLS_MODULUS, weight_adjustment]
+                )
+        r = add_polynomialcoeff(r, summand)
+    
+    return r
+```
+
+#### `vanishing_polynomialcoeff`
+
+```python
+def vanishing_polynomialcoeff(xs: Sequence[BLSFieldElement]) -> PolynomialCoeff:
+    """
+    Compute the vanishing polynomial on ``xs`` (in coefficient form)
+    """
+    p = [1]
+    for x in xs:
+        p = multiply_polynomialcoeff(p, [-int(x), 1])
+    return p
```

Comment:
(Thank @asn-d6 for discussion)

There is a bunch of `PolynomialCoeff` typing issues in many helper functions.

`PolynomialCoeff` is defined as `Vector[BLSFieldElement, FIELD_ELEMENTS_PER_BLOB]`, but the polynomialcoeff helpers are for various length of polynomials.

We could probably change `PolynomialCoeff` to a `List[BLSFieldElement, FIELD_ELEMENTS_PER_BLOB * 2]`. However, the Deneb `Polynomial` was defined as `Vector[BLSFieldElement, FIELD_ELEMENTS_PER_BLOB]`. 😭

Reply:
Indeed, `vanishing_polynomialcoeff()` uses `multiply_polynomialcoeff()` with variable degree polynomials, starting from degree 1 and going up to `len(xs)`. This is a problem because `multiply_polynomialcoeff()` and its subroutine `add_polynomialcoeff()`  take as input the static-sized `PolynomialCoeff` structure.

WRT spec readability I think there is value in having a stastic-sized structure representing the non-extended DAS poly of size `FIELD_ELEMENTS_PER_BLOB`, and maybe another structure representing an extended DAS polynomial of size `FIELD_ELEMENTS_PER_BLOB*2`.

WRT `multiply_polynomialcoeff()` taking variable-sized polynomials: we could perhaps have it work over a `Sequence[BLSFieldElement]` which would be correct but we would lose type safety/niceties.
Alternatively, we could define yet another polynomial structure called `VariableSizedPolynomial` or something which is `List[BLSFieldElement, FIELD_ELEMENTS_PER_BLOB * 2]`, but then casting from one polynomial type to another might get annoying.

Personally, I would merge the spec as-is, glossing over the current typing issues, and note them down as future work. In general, a less from 4844 is that as implementations move forward, we get more insight on what's the correct way to represent these polynomials.

Reply:
The vector was a mistake. I think `List[BLSFieldElement, FIELD_ELEMENTS_PER_BLOB * 2]` is the right definition, which I have fixed now.